operator/internal/handlers/internal/storage/secrets.go

```diff
@@ -440,16 +454,37 @@ func extractS3ConfigSecret(s *corev1.Secret, credentialMode lokiv1.CredentialMod
 		cfg.STS = true
 		cfg.Audience = string(audience)
 
+		err := validateS3Endpoint(endpoint, region)
+
 		// In the STS case region is not an optional field
 		if len(region) == 0 {
 			return nil, fmt.Errorf("%w: %s", errSecretMissingField, storage.KeyAWSRegion)
+		} else if len(endpoint) != 0 && err != nil {
+			return nil, err
 		}
 		return cfg, nil
 	default:
 		return nil, fmt.Errorf("%w: %s", errSecretUnknownCredentialMode, credentialMode)
 	}
 }
 
+func validateS3Endpoint(endpoint []byte, region []byte) error {
+	parsedURL, err := url.Parse(string(endpoint))
+	if err != nil || parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		return errS3UnparseableEndpoint
+	}
+	pattern := fmt.Sprintf(`s3.%s.amazonaws.com`, region)
+	matched, err := regexp.MatchString(pattern, string(endpoint))
+	if !matched {
+		return errS3InvalidEndpoint
+	}
+	if err != nil {
+		return errS3RegexpMatching
+	}
+
+	return nil
+}
```

Comment:
Some notes on the proposed changes
- Changed both parameters to string so we don't need to cast them every time
- Changed the validation flow to allow for more specific errors.
- Changed the error messages names so it's easier to read (this might only be for me though so feel free to take it or not)
```suggestion
const awsEndpoint = "amazonaws.com"

func validateS3Endpoint(endpoint string, region string) error {
	parsedURL, err := url.Parse(endpoint)
	if err != nil {
		return errS3EndpointUnparseable
	}
	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
		return errS3EndpointUnsuportedScheme
	}
	if !strings.HasSuffix(endpoint, awsEndpoint) {
		return errS3EndpointNotAWS
	}
	if !strings.HasSuffix(endpoint, fmt.Sprintf("%s.%s", region, awsEndpoint)) {
		return errS3EndpointWrongRegion
	}
	if !strings.HasSuffix(endpoint, fmt.Sprintf("s3.%s.%s", region, awsEndpoint)) {
		return errS3EndpointNotS3
	}
	return nil
}
```

Reply:
This is indeed more descriptive. Thanks!